File: Tests/PUPSISPortalTests/GridGeometryTests.swift
import XCTest
@testable import PUPSISPortal

final class GridGeometryTests: XCTestCase {
    /// 7 columns of 100pt with 5pt gaps = 730pt wide. 960pt tall over a
    /// 16-hour axis is one point per minute, which keeps the arithmetic here
    /// readable.
    private let grid = GridGeometry(
        width: 730,
        height: 960,
        axis: (start: 6 * 60, end: 22 * 60),
        columnSpacing: 5
    )

    func testColumnsDivideTheWidthEvenly() {
        XCTAssertEqual(grid.columnWidth, 100, accuracy: 0.001)
    }

    func testEachDayStartsWhereTheLastColumnEnded() {
        XCTAssertEqual(grid.x(of: .monday), 0, accuracy: 0.001)
        XCTAssertEqual(grid.x(of: .tuesday), 105, accuracy: 0.001)
        XCTAssertEqual(grid.x(of: .sunday), 630, accuracy: 0.001)
    }

    func testAPointResolvesToItsOwnColumn() {
        XCTAssertEqual(grid.day(atX: 0), .monday)
        XCTAssertEqual(grid.day(atX: 50), .monday)
        XCTAssertEqual(grid.day(atX: 110), .tuesday)
        XCTAssertEqual(grid.day(atX: 680), .sunday)
    }

    /// Dragging off either edge should keep selecting the end column rather
    /// than falling off the grid.
    func testPointsOutsideTheGridClampToTheEndColumns() {
        XCTAssertEqual(grid.day(atX: -500), .monday)
        XCTAssertEqual(grid.day(atX: 5000), .sunday)
    }

    func testAZeroWidthGridDoesNotDivideByZero() {
        let empty = GridGeometry(width: 0, height: 0, axis: (start: 0, end: 60))
        XCTAssertEqual(empty.day(atX: 40), .monday)
    }

    // MARK: Drag ranges

    func testASingleColumnDragCoversOneDay() {
        let range = grid.drag(from: CGPoint(x: 40, y: 480), to: CGPoint(x: 60, y: 600))

        XCTAssertEqual(range.days, [.monday])
        XCTAssertFalse(range.isMultiDay)
        XCTAssertEqual(range.start, 14 * 60)
        XCTAssertEqual(range.end, 16 * 60)
    }

    func testDraggingRightCoversEveryColumnBetween() {
        let range = grid.drag(from: CGPoint(x: 40, y: 480), to: CGPoint(x: 260, y: 600))

        XCTAssertEqual(range.days, [.monday, .tuesday, .wednesday])
        XCTAssertTrue(range.isMultiDay)
    }

    /// Dragging right-to-left has to mean the same days as left-to-right.
    func testDraggingLeftCoversTheSameDays() {
        let rightward = grid.drag(from: CGPoint(x: 40, y: 480), to: CGPoint(x: 260, y: 600))
        let leftward = grid.drag(from: CGPoint(x: 260, y: 600), to: CGPoint(x: 40, y: 480))

        XCTAssertEqual(leftward.days, rightward.days)
        XCTAssertEqual(leftward.start, rightward.start)
        XCTAssertEqual(leftward.end, rightward.end)
    }

    /// The event is stored starting on the first day; the rest come from the
    /// recurrence rule, so the anchor must be the earliest weekday.
    func testTheAnchorDayIsTheEarliestRegardlessOfDragDirection() {
        let leftward = grid.drag(from: CGPoint(x: 470, y: 480), to: CGPoint(x: 40, y: 600))

        XCTAssertEqual(leftward.anchorDay, .monday)
    }

    func testDragsAcrossTheWholeWeekCoverAllSevenDays() {
        let range = grid.drag(from: CGPoint(x: 0, y: 0), to: CGPoint(x: 730, y: 960))

        XCTAssertEqual(range.days, Weekday.allCases)
    }

    // MARK: Rects

    func testABlockRectMatchesItsColumnAndTimes() {
        let rect = grid.rect(day: .tuesday, start: 14 * 60, end: 16 * 60)

        XCTAssertEqual(rect.minX, 105, accuracy: 0.001)
        XCTAssertEqual(rect.width, 100, accuracy: 0.001)
        XCTAssertEqual(rect.minY, 480, accuracy: 0.001)
        XCTAssertEqual(rect.height, 120, accuracy: 0.001)
    }

    /// A zero-length block would give the editor popover nothing to anchor to.
    func testAZeroLengthBlockStillHasHeight() {
        let rect = grid.rect(day: .monday, start: 9 * 60, end: 9 * 60)

        XCTAssertGreaterThan(rect.height, 0)
    }

    func testRubberBandCoversTheColumnsItTouches() {
        let band = CGRect(x: 60, y: 100, width: 200, height: 300)

        XCTAssertEqual(grid.days(in: band), [.monday, .tuesday, .wednesday])
    }
}

File: windows/PUPSISPortal.Core.Tests/GridGeometryTests.cs
namespace PUPSISPortal.Core.Tests;

public class GridGeometryTests
{
    /// <summary>
    /// 7 columns of 100pt with 5pt gaps = 730pt wide. 960pt tall over a
    /// 16-hour axis is one point per minute, which keeps the arithmetic here
    /// readable.
    /// </summary>
    private readonly GridGeometry _grid = new(
        width: 730,
        height: 960,
        axis: (start: 6 * 60, end: 22 * 60),
        columnSpacing: 5
    );

    [Fact]
    public void TestColumnsDivideTheWidthEvenly()
    {
        Assert.Equal(100, _grid.ColumnWidth, precision: 3);
    }

    [Fact]
    public void TestEachDayStartsWhereTheLastColumnEnded()
    {
        Assert.Equal(0, _grid.X(Weekday.Monday), precision: 3);
        Assert.Equal(105, _grid.X(Weekday.Tuesday), precision: 3);
        Assert.Equal(630, _grid.X(Weekday.Sunday), precision: 3);
    }

    [Fact]
    public void TestAPointResolvesToItsOwnColumn()
    {
        Assert.Equal(Weekday.Monday, _grid.Day(0));
        Assert.Equal(Weekday.Monday, _grid.Day(50));
        Assert.Equal(Weekday.Tuesday, _grid.Day(110));
        Assert.Equal(Weekday.Sunday, _grid.Day(680));
    }

    /// <summary>
    /// Dragging off either edge should keep selecting the end column rather
    /// than falling off the grid.
    /// </summary>
    [Fact]
    public void TestPointsOutsideTheGridClampToTheEndColumns()
    {
        Assert.Equal(Weekday.Monday, _grid.Day(-500));
        Assert.Equal(Weekday.Sunday, _grid.Day(5000));
    }

    [Fact]
    public void TestAZeroWidthGridDoesNotDivideByZero()
    {
        var empty = new GridGeometry(width: 0, height: 0, axis: (start: 0, end: 60));
        Assert.Equal(Weekday.Monday, empty.Day(40));
    }

    // MARK: Drag ranges

    [Fact]
    public void TestASingleColumnDragCoversOneDay()
    {
        var range = _grid.Drag(originX: 40, originY: 480, currentX: 60, currentY: 600);

        Assert.Single(range.Days, Weekday.Monday);
        Assert.False(range.IsMultiDay);
        Assert.Equal(14 * 60, range.Start);
        Assert.Equal(16 * 60, range.End);
    }

    [Fact]
    public void TestDraggingRightCoversEveryColumnBetween()
    {
        var range = _grid.Drag(originX: 40, originY: 480, currentX: 260, currentY: 600);

        Assert.Equal(new[] { Weekday.Monday, Weekday.Tuesday, Weekday.Wednesday }, range.Days);
        Assert.True(range.IsMultiDay);
    }

    /// <summary>
    /// Dragging right-to-left has to mean the same days as left-to-right.
    /// </summary>
    [Fact]
    public void TestDraggingLeftCoversTheSameDays()
    {
        var rightward = _grid.Drag(originX: 40, originY: 480, currentX: 260, currentY: 600);
        var leftward = _grid.Drag(originX: 260, originY: 600, currentX: 40, currentY: 480);

        Assert.Equal(leftward.Days, rightward.Days);
        Assert.Equal(leftward.Start, rightward.Start);
        Assert.Equal(leftward.End, rightward.End);
    }

    /// <summary>
    /// The event is stored starting on the first day; the rest come from the
    /// recurrence rule, so the anchor must be the earliest weekday.
    /// </summary>
    [Fact]
    public void TestTheAnchorDayIsTheEarliestRegardlessOfDragDirection()
    {
        var leftward = _grid.Drag(originX: 470, originY: 480, currentX: 40, currentY: 600);

        Assert.Equal(Weekday.Monday, leftward.AnchorDay);
    }

    [Fact]
    public void TestDragsAcrossTheWholeWeekCoverAllSevenDays()
    {
        var range = _grid.Drag(originX: 0, originY: 0, currentX: 730, currentY: 960);

        var allDays = new[] { Weekday.Monday, Weekday.Tuesday, Weekday.Wednesday,
                              Weekday.Thursday, Weekday.Friday, Weekday.Saturday, Weekday.Sunday };
        Assert.Equal(allDays, range.Days);
    }

    // MARK: Rects

    [Fact]
    public void TestABlockRectMatchesItsColumnAndTimes()
    {
        var rect = _grid.Rect(Weekday.Tuesday, 14 * 60, 16 * 60);

        Assert.Equal(105, rect.MinX, precision: 3);
        Assert.Equal(100, rect.Width, precision: 3);
        Assert.Equal(480, rect.MinY, precision: 3);
        Assert.Equal(120, rect.Height, precision: 3);
    }

    /// <summary>
    /// A zero-length block would give the editor popover nothing to anchor to.
    /// </summary>
    [Fact]
    public void TestAZeroLengthBlockStillHasHeight()
    {
        var rect = _grid.Rect(Weekday.Monday, 9 * 60, 9 * 60);

        Assert.True(rect.Height > 0);
    }

    [Fact]
    public void TestRubberBandCoversTheColumnsItTouches()
    {
        var band = new GridRect(60, 100, 200, 300);

        var days = _grid.Days(band);
        Assert.Equal(new[] { Weekday.Monday, Weekday.Tuesday, Weekday.Wednesday }, days);
    }
}
